lib/raygun.breadcrumbs.ts

```diff
@@ -0,0 +1,139 @@
+import type { AsyncLocalStorage } from "async_hooks";
+import type { BreadcrumbMessage, Breadcrumb } from "./types";
+const debug = require("debug")("raygun");
+
+let asyncLocalStorage: AsyncLocalStorage<Breadcrumb[]> | null = null;
+
+try {
+  asyncLocalStorage = new (require("async_hooks").AsyncLocalStorage)();
+  debug("[raygun.breadcrumbs.ts] initialized successfully");
+} catch (e) {
+  debug(
+    "[raygun.breadcrumbs.ts] failed to load async_hooks.AsyncLocalStorage - initialization failed\n",
+    e,
+  );
+}
+
+type SourceFile = {
+  fileName: string;
+  functionName: string;
+  lineNumber: number | null;
+};
+
+function returnCallerSite(
+  err: Error,
+  callsites: NodeJS.CallSite[],
+): SourceFile | null {
+  for (const callsite of callsites) {
+    const fileName = callsite.getFileName() || "";
+
+    if (fileName.startsWith(__dirname)) {
+      continue;
+    }
+
+    return {
+      fileName: fileName,
+      functionName: callsite.getFunctionName() || "<anonymous>",
+      lineNumber: callsite.getLineNumber(),
+    };
+  }
+
+  return null;
+}
+
+function getCallsite(): SourceFile | null {
+  const originalPrepareStacktrace = Error.prepareStackTrace;
+
+  Error.prepareStackTrace = returnCallerSite;
+
+  // Ignore use of any, required for captureStackTrace
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const output: any = {};
+
+  Error.captureStackTrace(output);
+
+  const callsite = output.stack;
+  Error.prepareStackTrace = originalPrepareStacktrace;
+  return callsite;
+}
+
+export function addBreadcrumb(
+  breadcrumb: string | BreadcrumbMessage,
+  type: Breadcrumb["type"] = "manual",
+) {
+  const crumbs = getBreadcrumbs();
+
+  if (!crumbs) {
+    return;
+  }
+
+  if (typeof breadcrumb === "string") {
+    const expandedBreadcrumb: BreadcrumbMessage = {
+      message: breadcrumb,
+      level: "info",
+      category: "",
+    };
+
+    breadcrumb = expandedBreadcrumb;
+  }
+
+  const callsite = getCallsite();
+
+  const internalCrumb: Breadcrumb = {
+    ...(breadcrumb as BreadcrumbMessage),
+    category: breadcrumb.category || "",
+    message: breadcrumb.message || "",
+    level: breadcrumb.level || "info",
+    timestamp: Number(new Date()),
+    type,
+    className: callsite?.fileName,
+    methodName: callsite?.functionName,
+
+    // TODO - do we need to do any source mapping?
```

Comment:
Create issue for this?

Reply:
Done: https://github.com/MindscapeHQ/raygun4node/issues/224